AudioDeviceExample/AudioDevices/ExampleAVAudioEngineDevice.m

```diff
@@ -0,0 +1,886 @@
+            if (![self setupAudioEngine]) {
+                NSLog(@"Failed to setup AVAudioEngine");
+            }
+
+            // The manual rendering block (called in Core Audio's VoiceProcessingIO's playout callback at real time)
+            self.renderingContext->renderBlock = (__bridge void *)(_engine.manualRenderingBlock);
+
+            // Initialize the capturing context
+            self.capturingContext = malloc(sizeof(AudioCapturerContext));
+            memset(self.capturingContext, 0, sizeof(AudioCapturerContext));
+            self.capturingContext->bufferList = &_captureBufferList;
+        }
+
+        return self;
+    } else {
+        self = nil;
+        NSException *exception = [NSException exceptionWithName:@"ExampleAVAudioEngineDeviceNotSupported"
+                                                         reason:@"ExampleAVAudioEngineDevice requires iOS 11.0 or greater." userInfo:nil];
+        [exception raise];
+        return self;
+    }
+}
+
+- (void)dealloc {
+    [self unregisterAVAudioSessionObservers];
+
+    [self teardownAudioEngine];
+
+    free(self.renderingContext);
+    self.renderingContext = NULL;
+
+    free(self.capturingContext);
+    self.capturingContext = NULL;
+}
+
++ (NSString *)description {
+    return @"AVAudioEngine Audio Mixing";
+}
+
+/*
+ * Determine at runtime the maximum slice size used by VoiceProcessingIO. Setting the stream format and sample rate
+ * doesn't appear to impact the maximum size so we prefer to read this value once at initialization time.
+ */
++ (void)initialize {
+    AudioComponentDescription audioUnitDescription = [self audioUnitDescription];
+    AudioComponent audioComponent = AudioComponentFindNext(NULL, &audioUnitDescription);
+    AudioUnit audioUnit;
+    OSStatus status = AudioComponentInstanceNew(audioComponent, &audioUnit);
+    if (status != 0) {
+        NSLog(@"Could not find VoiceProcessingIO AudioComponent instance!");
+        return;
+    }
+
+    UInt32 framesPerSlice = 0;
+    UInt32 propertySize = sizeof(framesPerSlice);
+    status = AudioUnitGetProperty(audioUnit, kAudioUnitProperty_MaximumFramesPerSlice,
+                                  kAudioUnitScope_Global, kOutputBus,
+                                  &framesPerSlice, &propertySize);
+    if (status != 0) {
+        NSLog(@"Could not read VoiceProcessingIO AudioComponent instance!");
+        AudioComponentInstanceDispose(audioUnit);
+        return;
+    }
+
+    NSLog(@"This device uses a maximum slice size of %d frames.", (unsigned int)framesPerSlice);
+    kMaximumFramesPerBuffer = (size_t)framesPerSlice;
+    AudioComponentInstanceDispose(audioUnit);
+}
+
+#pragma mark - AudioEngine
+
+- (BOOL)setupAudioEngine {
+    NSAssert(_engine == nil, @"AVAudioEngine is already configured");
+
+    /*
+     * By default engine will render to/from the audio device make connections, e.g. inputNode -> effectNode -> outputNode
```

Comment:
Not sure quite what you were going for. Suggested text:

By default AVAudioEngine will render to/from the audio device, and automatically establish connections between nodes, e.g. inputNode -> effectNode -> outputNode.